### PR TITLE
Fix #72: feat(providers): add first-class smoke-test/health-check execution contracts for CLI providers

### LINKED ISSUE #72: feat(providers): add first-class smoke-test/health-check execution contracts for CLI providers
## Summary

Downstream apps still need to build provider-specific test commands just to verify that a CLI-backed provider is installed, authenticated, and minimally responsive.

In Paid, `Providers::TestAgent` currently owns a second provider command map plus custom smoke-test wrappers for Codex, Gemini, Kilocode, OpenCode, and Copilot-adjacent behavior.

## Why This Should Move Upstream

A provider's smoke-test contract is part of its adapter contract, not app-specific behavior.

If `agent-harness` already knows how to run a provider for real execution, it should also be able to describe or perform a minimal provider-specific health/smoke test that downstream apps can reuse for:
- installation verification
- auth verification
- minimal responsiveness checks
- stable error classification for common failure modes

Without that, downstream apps duplicate provider-specific CLI knowledge in a second code path that drifts separately from normal execution.

## Current Downstream Consequence

Paid currently has to maintain local smoke-test behavior for multiple providers, including:
- provider-specific ping command construction
- shell wrappers to capture last-message output or error reports
- provider-specific error extraction/classification

## Expected Behavior

`agent-harness` should expose a first-class provider smoke-test / health-check contract for CLI-backed providers.

That could be:
- a `check_provider` path that can execute the real provider-specific smoke test through the configured executor
- provider metadata describing a canonical smoke-test command/prompt/expectation
- or a higher-level `smoke_test` API that returns normalized status/error information

## Acceptance Criteria

- CLI-backed providers can expose or execute a canonical smoke-test contract through `agent-harness`.
- Downstream apps no longer need to maintain a separate provider command map just for health checks.
- Smoke-test failures are normalized enough for downstream apps to distinguish installation/auth/rate-limit/timeout failures without provider-specific shell parsing where possible.
- The contract works with custom/per-request executors so containerized apps can use it.

## Non-Goals

This is not about full run orchestration or installer ownership. Those are tracked separately.

## Related

- Related to viamin/paid#784
- Related to viamin/agent-harness#60
- Related to viamin/agent-harness#61
- Related to viamin/agent-harness#62

### PR DESCRIPTION
**Pull request by viamin**
Implemented the provider-owned smoke-test contract and committed it as `ea2d1cd` (`feat(providers): add cli smoke test contracts`).

The main change is that CLI-backed providers now expose a first-class smoke-test contract in the adapter/base layer, and [`ProviderHealthCheck`](/workspace/lib/agent_harness/provider_health_check.rb) now executes that canonical smoke test through the selected executor instead of stopping at registration/auth/config checks. Health-check results now include normalized `error_category` and `check` metadata so downstream callers can distinguish installation, authentication, configuration, timeout, rate-limit, and related failure modes without provider-specific shell parsing. I also exposed contract lookup on the public API and registry in [`lib/agent_harness.rb`](/workspace/lib/agent_harness.rb) and [`lib/agent_harness/providers/registry.rb`](/workspace/lib/agent_harness/providers/registry.rb).

Verification passed end to end: `bundle exec rubocop`, `bundle exec rspec`, and the git commit hooks all succeeded. The worktree is clean.

---

Closes #72

### CODE REVIEW

**Review by viamin — COMMENTED** [review]
I found two correctness issues in the new smoke-test path. The first is that `Providers::Base` now opts every provider into the same generic contract, which is not valid for provider CLIs whose execution mode is not free-form chat (for example, GitHub Copilot's `what-the-shell`). The second is that `Adapter#smoke_test` treats any successful non-empty output as a pass, so the health check cannot verify the provider actually satisfied the contract's expected response.

File: lib/agent_harness/provider_health_check.rb [hunk 1]
```diff
@@ -194,14 +214,16 @@ def perform_check(provider_name, start_time)
         # The Adapter default always returns {healthy: true}, so providers
         # that haven't implemented a real health check are reported as ok
         # with a note that the check is not implemented.
-        provider_instance = build_provider(provider_name, klass)
+        provider_instance = build_provider(provider_name, klass, executor: executor)
```
Comment (viamin) [inline lib/agent_harness/provider_health_check.rb]: Forwarding the custom `executor:` only here still leaves the auth preflight on the host path. `perform_check` calls `Authentication.auth_status(provider_name)` before this, and `Authentication.resolve_provider` always builds that provider with `AgentHarness.configuration.command_executor` plus host env/files. For providers like Codex, a Docker executor with a container-only `OPENAI_API_KEY` will still fail authentication before the smoke test ever runs, so the PR does not actually satisfy issue #72's custom-executor acceptance criterion.

File: lib/agent_harness/provider_health_check.rb [hunk 2]
```diff
@@ -216,7 +238,42 @@ def perform_check(provider_name, start_time)
...
+        end
+
+        smoke = provider_instance.smoke_test(timeout: timeout, provider_runtime: provider_runtime)
+        unless smoke[:ok]
+          return build_result(
+            name: provider_name,
+            status: smoke[:status] || "error",
+            message: smoke[:message] || "Smoke test failed",
+            start_time: start_time,
+            error_category: smoke[:error_category] || :unknown,
```
Comment (viamin) [inline lib/agent_harness/provider_health_check.rb]: `ProviderHealthCheck` now exposes two different `error_category` vocabularies. The preflight branches emit health-check categories like `:installation` and `:authentication`, but the smoke-test branch forwards adapter/ErrorTaxonomy categories like `:auth_expired` and `:rate_limited` unchanged. That means the same auth failure is reported as `:authentication` when it happens in `auth_status`, and `:auth_expired` when it happens during the smoke test. Downstream callers still have to special-case provider internals, which is the opposite of the normalization promised in #72 and in this PR description. Please normalize smoke-test categories at the health-check boundary before returning them here.

**Review by viamin — COMMENTED** [review]
I found two blocking issues in the new smoke-test health-check path. Both matter to the linked #72 acceptance criteria: custom executors/provider runtime still do not cover the full auth path, and the new `error_category` field now mixes two different taxonomies depending on which branch fails. I was not able to run the targeted specs in this environment because `bundle exec rspec` fails here with `bundler: command not found: rspec`.

**Review by viamin — COMMENTED** [review]
Issue #72 is largely addressed: the PR adds provider-owned smoke-test contracts, routes health checks through the provider executor path, and exposes the contract through the public registry/API. I found one regression in the new execution-context gating logic that can hide installation/auth failures when callers pass an empty runtime payload.

I couldn't run the targeted RSpec files in this environment because `bundle exec rspec` fails here with `bundler: command not found: rspec`.

**Comment by viamin** [comment]
## Agent Update

Updated the unresolved review thread in [provider_health_check.rb](/workspace/lib/agent_harness/provider_health_check.rb#L333) so host preflight still runs when `provider_runtime` is present but empty. `host_preflight_allowed?` now treats `nil` and `ProviderRuntime.wrap(...).empty?` equivalently when no custom executor is supplied, which fixes the false-positive `ok` path the reviewer reported for `check(:provider, provider_runtime: {})`.

I added a regression spec in [provider_health_check_spec.rb](/workspace/spec/agent_harness/provider_health_check_spec.rb#L55) covering the exact case: a provider with `available? == false` still returns the installation/availability failure when called with an empty runtime hash.

Verification passed:
- `bundle install --path vendor/bundle`
- `bundle exec rubocop`
- `bundle exec rspec`
- `git commit` hooks

Commit: `1d03721` (`fix(health-check): preserve host preflight for empty runtimes`)  

Worktree is clean.

File: lib/agent_harness/provider_health_check.rb [hunk 3]
```diff
@@ -258,25 +330,65 @@ def auth_not_implemented?(auth)
         error.include?("not implemented")
       end
 
+      def host_preflight_allowed?(executor:, provider_runtime:)
+        return false unless executor.nil?
+        return true if provider_runtime.nil?
+
+        ProviderRuntime.wrap(provider_runtime).empty?
```
Comment (viamin) [inline lib/agent_harness/provider_health_check.rb]: `provider_runtime` is not a reliable signal that execution moved off-host. It also carries ordinary per-request overrides like `model`, `flags`, `base_url`, and `env` for the local executor. With this gate, `check(:provider, provider_runtime: {model: "..."})` skips availability/auth/health preflight entirely and can return `ok` even when the CLI is not installed. I reproduced that locally with a provider whose `available?` returns `false`: the plain check returns the expected installation error, but the same check with only `provider_runtime: {model: "runtime-only"}` returns `"Smoke test passed using the supplied execution context"`.

```suggestion
      def host_preflight_allowed?(executor:, provider_runtime:)
        return false unless executor.nil?

        true
      end
```

File: lib/agent_harness/provider_health_check.rb [hunk 4]
```diff
@@ -25,35 +25,45 @@ class << self
...
+      def check_all(timeout: configured_timeout, executor: nil, provider_runtime: nil)
         provider_names = if AgentHarness.configuration.providers.empty?
           Providers::Registry.instance.all
         else
           enabled_provider_names
         end
 
-        provider_names.map { |name| check(name, timeout: timeout) }
+        provider_names.map do |name|
+          check(name, timeout: timeout, executor: executor, provider_runtime: provider_runtime)
```
Comment (viamin) [inline lib/agent_harness/provider_health_check.rb]: `check_all` now forwards a single `provider_runtime` object to every enabled provider. Since `ProviderRuntime` can include provider-specific `env`, `flags`, `base_url`, and `metadata`, calling `AgentHarness.check_providers(provider_runtime: ...)` will bleed one provider's runtime overrides into every other provider health check. That is especially risky for credentials in `env`. I think runtime overrides should stay on `check_provider`, or this API needs a per-provider map instead of one shared object.

**Review by viamin — COMMENTED** [review]
I found two blocking regressions in the new health-check contract flow. The main one is that any non-empty `provider_runtime` now disables host preflight even when execution is still local, so `check_provider(..., provider_runtime: {model: ...})` can report success while the CLI is missing. The second is that `check_all`/`check_providers` applies one runtime override to every provider, which can leak provider-specific env/flags/base URLs across unrelated health checks.

**Comment by viamin** [comment]
## Agent Update

Addressed both unresolved review threads.

`provider_runtime` no longer disables host preflight by itself in [provider_health_check.rb](/workspace/lib/agent_harness/provider_health_check.rb). Health checks still verify local CLI availability/auth/health/config unless a custom `executor` is supplied, which fixes the false-positive `ok` path for local runtime-only overrides.

Bulk health checks now reject shared runtime overrides explicitly in [provider_health_check.rb](/workspace/lib/agent_harness/provider_health_check.rb) and [agent_harness.rb](/workspace/lib/agent_harness.rb), so `check_all` / `AgentHarness.check_providers` cannot leak one provider’s `env`, `flags`, or other runtime settings into every other provider. Regression coverage was added in [provider_health_check_spec.rb](/workspace/spec/agent_harness/provider_health_check_spec.rb).


Verification passed with `bundle install --path vendor/bundle`, `bundle exec rubocop`, `bundle exec rspec`, and the commit hook during `git commit`. Commit: `fc9e282` (`fix(health-check): preserve host preflight and reject shared runtimes`). Worktree is clean.

**Review by viamin — COMMENTED** [review]
I found two issues worth addressing before this is considered complete: one new smoke-test failure mode is now harder to diagnose, and the linked issue still looks only partially covered because the Copilot CLI path is not wired into the new contract.

**Comment by viamin** [comment]
## Agent Update

Addressed both unresolved review threads and committed the changes as `5fca178` (`fix(providers): complete smoke test review follow-ups`).

In [provider_health_check.rb](/workspace/lib/agent_harness/provider_health_check.rb), `ConfigurationError` now follows the same explicit path as `NotImplementedError`, so malformed smoke-test contracts keep their real diagnostic message and are categorized as `:configuration` instead of being collapsed into a generic unknown failure. In [github_copilot.rb](/workspace/lib/agent_harness/providers/github_copilot.rb), `GithubCopilot` now publishes the canonical first-class smoke-test contract, closing the remaining CLI provider gap the review called out.

I added regression coverage in [provider_health_check_spec.rb](/workspace/spec/agent_harness/provider_health_check_spec.rb), [registry_spec.rb](/workspace/spec/agent_harness/providers/registry_spec.rb), and [github_copilot_spec.rb](/workspace/spec/agent_harness/providers/github_copilot_spec.rb). I also reviewed the touched diff for the same class of issue and confirmed all built-in CLI providers registered in the registry now expose smoke-test contracts.

Verification passed:
- `bundle exec rubocop`
- `bundle exec rspec`
- `git commit` hooks

The worktree is clean.

File: lib/agent_harness/provider_health_check.rb [hunk 5]
```diff
@@ -258,25 +332,62 @@ def auth_not_implemented?(auth)
         error.include?("not implemented")
       end
 
+      def host_preflight_allowed?(executor:, provider_runtime: nil)
```
Comment (viamin) [inline lib/agent_harness/provider_health_check.rb]: This only skips host preflight when the caller passes `executor:` directly. If an app configures a non-host executor globally (for example `configuration.command_executor = DockerCommandExecutor.new(...)`), `build_provider` will execute the smoke test in that container, but `klass.available?` and `Authentication.auth_status` still run on the host and can fail first. That means the PR still returns false negatives for one of the execution-context paths that #72 explicitly calls out.

```suggestion
      def host_preflight_allowed?(executor:, provider_runtime: nil)
        effective_executor = executor || AgentHarness.configuration.command_executor
        effective_executor.instance_of?(CommandExecutor)
      end
```

**Review by viamin — COMMENTED** [review]
The new smoke-test API and normalized result shape line up with issue #72, but I found one blocker in the execution-context handling: a globally configured non-local executor still goes through host-side availability/auth preflight, so health checks can fail before the smoke test runs in the configured context.

**Comment by viamin** [comment]
## Agent Update

Updated the health-check executor gate in [provider_health_check.rb](/workspace/lib/agent_harness/provider_health_check.rb) so host preflight now follows the effective executor, including `AgentHarness.configuration.command_executor`, instead of only the explicit `executor:` argument. That addresses the unresolved review comment: a globally configured non-host executor now skips host-only availability/auth checks and runs the provider smoke test in the configured execution context.

I added a regression in [provider_health_check_spec.rb](/workspace/spec/agent_harness/provider_health_check_spec.rb) covering the exact global non-host executor path, and reviewed the local diff for the same explicit-vs-effective configuration bug class before committing.


Verification passed with `bundle install --path vendor/bundle`, `bundle exec rubocop`, `bundle exec rspec`, and the git commit hook. Commit: `730b29b` (`fix(health-check): respect configured executor context`). Worktree is clean.

File: lib/agent_harness/providers/github_copilot.rb [hunk 6]
```diff
@@ -56,6 +56,10 @@ def discover_models
           ]
         end
 
+        def smoke_test_contract
+          Base::DEFAULT_SMOKE_TEST_CONTRACT
```
Comment (viamin) [inline lib/agent_harness/providers/github_copilot.rb]: `GithubCopilot`'s existing `send_message` path is hard-wired to `github-copilot-cli what-the-shell`, so the new default smoke-test prompt (`Reply with exactly OK.`) is not exercising a chat-style round trip here. In practice this is much more likely to return a shell suggestion than the literal `OK`, which turns health checks into false negatives for an otherwise healthy provider. Until this provider has a custom smoke test, I think it should opt out of the generic contract.

```suggestion
        def smoke_test_contract
          nil
        end
```

File: lib/agent_harness/provider_health_check.rb [hunk 7]
```diff
@@ -258,25 +332,63 @@ def auth_not_implemented?(auth)
...
+        effective_executor = executor || AgentHarness.configuration.command_executor
+        effective_executor.instance_of?(CommandExecutor)
```
Comment (viamin) [inline lib/agent_harness/provider_health_check.rb]: Using `instance_of?(CommandExecutor)` is too strict here: any local executor subclass (for example, a wrapper that adds logging, metrics, retries, or environment isolation while still executing on the host) now gets treated as a remote execution context and skips availability/auth preflight. That is a regression from the previous behavior for host-backed executors. Matching on `is_a?(CommandExecutor)` and explicitly excluding the known remote executor keeps subclasses working.

```suggestion
      def host_preflight_allowed?(executor:, provider_runtime: nil)
        effective_executor = executor || AgentHarness.configuration.command_executor
        effective_executor.is_a?(CommandExecutor) && !effective_executor.is_a?(DockerCommandExecutor)
      end
```

**Review by viamin — COMMENTED** [review]
I found two correctness issues in the new smoke-test path. One is provider-specific (`GithubCopilot` advertises the default exact-`OK` contract even though its command path is `what-the-shell`), and the other is in executor detection (`instance_of?(CommandExecutor)` now skips host preflight for any local executor subclass).

**Comment by viamin** [comment]
## Agent Update

Addressed both unresolved review threads and committed the fix as `dfb93ae` (`fix(providers): resolve smoke test review threads`).

`GithubCopilot` now opts out of the generic smoke-test contract in [github_copilot.rb](/workspace/lib/agent_harness/providers/github_copilot.rb), which avoids a false-negative health check on its `what-the-shell` command path. I updated the public contract expectations accordingly in [github_copilot_spec.rb](/workspace/spec/agent_harness/providers/github_copilot_spec.rb) and [registry_spec.rb](/workspace/spec/agent_harness/providers/registry_spec.rb).

`ProviderHealthCheck` now treats local `CommandExecutor` subclasses as host-backed while still excluding `DockerCommandExecutor` in [provider_health_check.rb](/workspace/lib/agent_harness/provider_health_check.rb). I added a regression for that exact subclass case, and adjusted the non-host executor coverage to use a real Docker executor instance, in [provider_health_check_spec.rb](/workspace/spec/agent_harness/provider_health_check_spec.rb).

Verification passed: `bundle install --path vendor/bundle`, `bundle exec rubocop`, `bundle exec rspec`, and the git commit hook. The worktree is clean.

**Review by viamin — COMMENTED** [review]
This mostly lines up with issue #72, but there is one blocking correctness issue around executor overrides.

File: lib/agent_harness/provider_health_check.rb [hunk 8]
```diff
@@ -216,7 +244,46 @@ def perform_check(provider_name, start_time)
...
             status: "degraded",
             message: "Configuration issues: #{errors_msg}",
-            start_time: start_time
+            start_time: start_time,
+            error_category: :configuration,
+            check: :configuration
+          )
+        end
+
+        smoke_contract = provider_instance.smoke_test_contract
```
Comment (viamin) [inline lib/agent_harness/provider_health_check.rb]: Consider handling the case where `provider_instance.smoke_test_contract` returns nil more explicitly; currently the fallback logic may mask missing contracts.

**Review by viamin — COMMENTED** [review]
Overall review: good addition of smoke-test contracts, consistent with project style.

**Review by viamin — COMMENTED** [review]
Overall review: The new smoke-test contract methods are a valuable addition and align with the feature request. Code follows project style and YARD documentation. Minor suggestion: ensure the method raises ConfigurationError explicitly if provider not found, or document that the registry will raise.

**Review by viamin — COMMENTED** [review]
Overall, the PR correctly introduces smoke‑test contract methods without affecting existing functionality. The implementation follows existing patterns, includes proper error handling, and respects code style conventions.

File: lib/agent_harness.rb [hunk 9]
```diff
@@ -117,6 +117,22 @@ def installation_contracts
       Providers::Registry.instance.installation_contracts
     end
 
+    # Get smoke-test metadata for a provider CLI.
```
Comment (viamin) [inline lib/agent_harness.rb]: The new `smoke_test_contract` method validates registration before delegating to the registry, which is consistent with other contract methods. Consider adding a thin wrapper similar to `provider_install_contract` for symmetry, but not required.

**Review by viamin — COMMENTED** [review]
I found two functional issues in the new smoke-test health-check path: request-scoped runtime overrides still execute host auth preflight, and github_copilot remains permanently degraded because it opts out of the new contract.

**Review by viamin — COMMENTED** [review]
I found two issues worth addressing before this can be considered complete for #72. The main one is that the new custom-executor path still calls provider-specific health/config hooks that read global state, so built-ins like Codex and Gemini can return false negatives even when the supplied executor/runtime is valid. The second is issue linkage: GitHub Copilot still explicitly opts out of a first-class smoke-test contract, which means downstream code still needs provider-specific health-check behavior for at least one CLI provider.

**Review by viamin — COMMENTED** [review]
I found two blocking issues in the new smoke-test flow. One is a behavioral regression around request-scoped runtime overrides, and the other is that GitHub Copilot still does not expose an upstream smoke-test contract even though issue #72 explicitly calls out Copilot-specific downstream duplication.

**Review by viamin — COMMENTED** [review]
## PR Review: Smoke Test/Health Check Implementation

This PR successfully implements first-class smoke-test execution contracts for CLI providers as requested in issue #72. Overall, this is a well-designed and comprehensive implementation with excellent test coverage.

### ✅ Strengths

**Architecture & Design:**
- Clean separation of concerns with provider-owned smoke-test contracts
- Comprehensive error taxonomy with normalized error categories
- Execution context awareness (host vs container environments)
- Integration with existing authentication and health check systems

**Implementation Quality:**
- All 9 built-in providers implement smoke test contracts consistently
- Default smoke test contract is sensible and minimally invasive
- Proper timeout handling and error classification
- Good backward compatibility with existing provider interface

**Test Coverage:**
- Excellent test coverage with 1,253+ lines in provider_health_check_spec.rb
- Comprehensive scenario testing including edge cases
- Tests for all major health check phases and error conditions
- Proper integration testing with orchestration layer

**API Design:**
- Clean public API with `AgentHarness.smoke_test_contract()` and `AgentHarness.check_provider()`
- Consistent result format with error categorization
- Support for custom executors and runtime contexts

### 🔍 Considerations

**Performance:**
- Default 30-second timeout for smoke tests seems reasonable for minimal health checks
- Provider registration and availability checks are efficient
- No obvious N+1 query issues or unnecessary allocations

**Security:**
- Good error message sanitization to avoid leaking sensitive information
- Proper handling of authentication and credential errors
- Secure by default with minimal exposure of internal errors

**Code Quality:**
- Well-structured code following existing conventions
- Proper error handling with specific exception types
- Good use of Ruby idioms and patterns

### 🎯 Verification

The implementation successfully addresses the acceptance criteria from issue #72:

✅ CLI-backed providers expose canonical smoke-test contracts  
✅ Downstream apps can reuse provider-specific health checks  
✅ Smoke-test failures are normalized with proper error categorization  
✅ Works with custom executors for containerized environments  

### 📋 Summary

This is a high-quality implementation that successfully addresses the requirements. The code is well-tested, follows best practices, and provides a solid foundation for provider health monitoring. The architecture is extensible and the API is clean and intuitive.

**Recommendation: APPROVE** - This PR is ready for merge.

The implementation effectively eliminates the need for downstream apps to maintain separate provider command maps for health checks, achieving the primary goal of moving provider-specific smoke test knowledge upstream into agent-harness.

**Comment by viamin** [comment]
## Agent Update

Commit succeeded. Here's what I did to address the review comment:

**Problem:** `perform_check` always forwarded the health-check timeout (default 5s) into `smoke_test`, overriding the provider contract's own timeout (30s for built-ins). This caused false-negative timeout failures for network-backed providers.

**Fix (3 changes):**

1. **`provider_health_check.rb:283`** — Pass `nil` to `smoke_test` instead of the health-check timeout, so `smoke_test` falls through to `contract[:timeout]` (adapter.rb:320: `timeout: timeout || contract[:timeout]`)

2. **`provider_health_check.rb:54`** — Extend the outer `Timeout.timeout` to `max(health_check_timeout, contract_timeout)` via new `effective_check_timeout` helper, so the contract timeout isn't killed by the shorter health-check timeout

3. **`provider_health_check.rb:351-361`** — New `effective_check_timeout` private method that looks up the provider's contract timeout from the registry and returns the larger of the two timeouts

**Tests added/updated:**
- New context "when provider contract timeout exceeds health-check timeout" with two specs: verifies `nil` is passed to `smoke_test`, and verifies outer timeout is extended to 45s
- Updated existing timeout test to expect 30s (contract timeout) instead of 2s (health-check timeout) for registered providers
- Added test for unregistered providers still using the caller timeout

**Comment by viamin** [comment]
## Agent Update

Commit `85f96cf` succeeded. Pre-commit hooks (rubocop + tests) passed.

**Summary of changes:**

- **`provider_health_check.rb`**: Added a `klass.available?` gate (Step 2a) before the executor-level `which` check (Step 2b) in the CLI preflight path. This honors custom providers that return `false` from `.available?` for reasons beyond PATH presence (e.g., version/feature requirements). The gate is only applied when no explicit executor is supplied, since `.available?` always queries the global executor.

- **`provider_health_check_spec.rb`**: Updated existing tests to expect the new `.available?` error message, and added a new test for the edge case where `.available?` returns `true` but the executor can't find the binary (verifying Step 2b still works independently).

**Comment by viamin** [comment]
## Agent Update

Commit `f9914e1` succeeded. Pre-commit hooks (rubocop + tests) passed.

**Fix:** In `provider_health_check.rb:302-306`, `smoke_test` was always called with `timeout: nil`. Now it conditionally forwards the caller's validated timeout when no contract is available:

- **With contract** (`smoke_contract` present): passes `nil` so the adapter falls through to `contract[:timeout]` — no change in behavior
- **Without contract** (provider overrides `#smoke_test` directly): passes the health-check `timeout` so the override can respect it instead of running unbounded

Updated the corresponding spec to assert the timeout is forwarded (`eq(9)`) for the custom-executor + no-contract case.
